PublicApiAnalyzer refinements

tobias-tengler opened the pull request:
- Enforce documentation of public API only for PRs targeting the `main` branch
- Separate pipeline job/Nuke target for checking the public API
- Convert powershell scripts to Nuke
- Exclude PublicAPI files from search results
- Add new APIs to unshipped

michaelstaib commented:
I will have a look.

michaelstaib commented:
I will do this tomorrow. :)

tobias-tengler commented on .build/Build.CheckPublicApi.cs:
```diff
@@ -0,0 +1,27 @@
+
+partial class Build : NukeBuild
+{
+    Target CheckPublicApi => _ => _
+        .DependsOn(Restore)
+        .Executes(() =>
+        {
+            if (!InvokedTargets.Contains(Restore))
+            {
+                DotNetBuildSonarSolution(AllSolutionFile);
+            }
+
+            DotNetBuild(c => c
+                .SetProjectFile(AllSolutionFile)
+                .SetNoRestore(InvokedTargets.Contains(Restore))
+                .SetConfiguration(Configuration)
+                .SetProperty("RequireDocumentationOfPublicApiChanges", true));
+        });
```
@michaelstaib I've added a new Nuke target, so the public API check can be executed via
```sh
./build.sh CheckPublicApi
```
All that's left is that there needs to be a PR check that runs when the PR targets main and the tests have successfully completed.

PascalSenn replied:
@tobias-tengler  This can be found here:
https://github.com/ChilliCream/hotchocolate/blob/main/.devops/azure-pipelines.test-pr-hotchocolate.yml

tobias-tengler replied:
@PascalSenn I've integrated it in the pipeline successfully!
But there's still a weird issue: I intentionally introduced an undocumented change and locally `./build.sh CheckPublicApi` errors out as expected, but in CI it passes 🤔
Not sure what's going on yet, but I'm on it...

tobias-tengler commented:
> SonarCloud Quality Gate failed.
> 
> [<img alt="Bug" width="16" height="16" src="https://camo.githubusercontent.com/deaad8c9d9b09977da3f9aa455689440a17356bb359560328463022a0d9d1362/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f6275672e706e67">](https://sonarcloud.io/project/issues?id=HotChocolate&pullRequest=3677&resolved=false&types=BUG) [<img alt="A" width="16" height="16" src="https://camo.githubusercontent.com/c51bc537468c04ce22bf2a646097b829e435b6847a82d6e663dd26de8ba24fc5/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f412e706e67">](https://sonarcloud.io/project/issues?id=HotChocolate&pullRequest=3677&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=HotChocolate&pullRequest=3677&resolved=false&types=BUG)
> [<img alt="Vulnerability" width="16" height="16" src="https://camo.githubusercontent.com/60273222b26453a76e9c369287fa0106b151cbaae4fa4ce00e149b269a0fac67/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f76756c6e65726162696c6974792e706e67">](https://sonarcloud.io/project/issues?id=HotChocolate&pullRequest=3677&resolved=false&types=VULNERABILITY) [<img alt="A" width="16" height="16" src="https://camo.githubusercontent.com/c51bc537468c04ce22bf2a646097b829e435b6847a82d6e663dd26de8ba24fc5/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f412e706e67">](https://sonarcloud.io/project/issues?id=HotChocolate&pullRequest=3677&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=HotChocolate&pullRequest=3677&resolved=false&types=VULNERABILITY)
> [<img alt="Security Hotspot" width="16" height="16" src="https://camo.githubusercontent.com/a0723b4c04ec36cc8d010cca748afdc6300592f020130edca7403ea5b17ffe35/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f73656375726974795f686f7473706f742e706e67">](https://sonarcloud.io/project/security_hotspots?id=HotChocolate&pullRequest=3677&resolved=false&types=SECURITY_HOTSPOT) [<img alt="A" width="16" height="16" src="https://camo.githubusercontent.com/c51bc537468c04ce22bf2a646097b829e435b6847a82d6e663dd26de8ba24fc5/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f412e706e67">](https://sonarcloud.io/project/security_hotspots?id=HotChocolate&pullRequest=3677&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=HotChocolate&pullRequest=3677&resolved=false&types=SECURITY_HOTSPOT)
> [<img alt="Code Smell" width="16" height="16" src="https://camo.githubusercontent.com/1b640677486022b878fa0367aa2b5cffad9b009d2ff2fc3440cbd7f8da85fe55/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636f6d6d6f6e2f636f64655f736d656c6c2e706e67">](https://sonarcloud.io/project/issues?id=HotChocolate&pullRequest=3677&resolved=false&types=CODE_SMELL) [<img alt="C" width="16" height="16" src="https://camo.githubusercontent.com/46f32ae0a124fed609facb7b9d8e97c97222c470eb8a1a0b951f5e07ef7f37c4/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f526174696e6742616467652f432e706e67">](https://sonarcloud.io/project/issues?id=HotChocolate&pullRequest=3677&resolved=false&types=CODE_SMELL) [1 Code Smell](https://sonarcloud.io/project/issues?id=HotChocolate&pullRequest=3677&resolved=false&types=CODE_SMELL)
> 
> [<img alt="100.0%" width="16" height="16" src="https://camo.githubusercontent.com/09ce026c321b1f22435cfe26c44012ab2cd647042a3300a035d6df6589b3cac8/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f436f76657261676543686172742f3130302e706e67">](https://sonarcloud.io/component_measures?id=HotChocolate&pullRequest=3677&metric=new_coverage&view=list) [100.0% Coverage](https://sonarcloud.io/component_measures?id=HotChocolate&pullRequest=3677&metric=new_coverage&view=list)
> [<img alt="0.0%" width="16" height="16" src="https://camo.githubusercontent.com/17c8580971832aaa80226d5b1c5c61cc2f9ef6a863fc6f6bae0ab2c7d45165ed/68747470733a2f2f736f6e6172736f757263652e6769746875622e696f2f736f6e6172636c6f75642d6769746875622d7374617469632d7265736f75726365732f76322f636865636b732f4475706c69636174696f6e732f332e706e67">](https://sonarcloud.io/component_measures?id=HotChocolate&pullRequest=3677&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=HotChocolate&pullRequest=3677&metric=new_duplicated_lines_density&view=list)

This Code-Smell should be ignored. It's that the parameter name on the implementation is different from the one on the interface. I changed this in another PR, but technically it would've been a breaking change. I reverted it and essentially just undid my previous change:

https://github.com/ChilliCream/hotchocolate/pull/3677/files#diff-da4e9423d08f59f3dd6a29541a5b52914417755475add15c36627fcecaa80a9dL116-R117

michaelstaib commented on .devops/azure-pipelines.test-pr-hotchocolate.yml:
```diff
@@ -34,3 +34,16 @@ stages:
             inputs:
               artifactName: test-results
               pathtoPublish: "output/test-results"
+      - job: CheckPublicApi
```
Lets spin this of in a GitHub action that is specifically for this task.

tobias-tengler replied:
I created one, but I can't seem to be able to test it with `act` locally. Is it possible to run the action in this PR already?

tobias-tengler commented:
@michaelstaib Thanks for updating the nuke build!
I'll update the documentation and remove the obsolete scripts. How can I specify the `From` and `To` parameters on the command line? I tried `--from main` or `From=main`, but it didn't work.

tobias-tengler commented:
@michaelstaib I added the remaining Nuke targets and also tested the GitHub action in my fork of HotChocolate.
This PR should be good to go! :rocket:

tobias-tengler commented:
@michaelstaib I'm not exactly sure how the changes here have messed with the build, but it seems like we should install all the frameworks we're targeting in the build pipeline, e.g. `net5.0` and `netcoreapp3.1`.
Should I do this or do you know something else that could be the cause?

michaelstaib commented:
> This Code Smell should be ignored. It's that the parameter name on the implementation is different from the one on the interface. I removed this Code Smell in another PR, but technically it's a breaking change. Therefor I reverted it and essentially just undid my previous change:

we can ignore this since the API is obsolete and will be removed in future. I will override the validation.

michaelstaib commented:
Build seems to work now again. I updated the devops pipelines to also install the old 3.1 framework.

michaelstaib commented:
The red tests are flaky ... will deal with them in a separate issue.
